crates/chain/tests/test_tx_graph.rs

```diff
@@ -1172,3 +1174,71 @@ fn test_missing_blocks() {
         ),
     ]);
 }
+
+#[test]
+/// The `map_anchors` allow a caller to pass a function to reconstruct the [`TxGraph`] with any [`Anchor`],
+/// even though the function is non-deterministic.
+fn call_map_anchors_with_non_deterministic_anchor() {
+    #[derive(Debug, Default, Clone, PartialEq, Eq, Copy, PartialOrd, Ord, core::hash::Hash)]
+    /// A non-deterministic anchor
+    pub struct NonDeterministicAnchor {
+        pub anchor_block: BlockId,
+        pub non_deterministic_field: u32,
+    }
+
+    let template = [
+        TxTemplate {
+            tx_name: "genesis",
+            inputs: &[TxInTemplate::Bogus],
+            outputs: &[TxOutTemplate::new(10000, Some(1))],
+            anchors: &[block_id!(1, "A")],
+            last_seen: None,
+        },
+        TxTemplate {
+            tx_name: "block 1",
+            inputs: &[TxInTemplate::PrevTx("genesis", 0)],
+            outputs: &[TxOutTemplate::new(20000, Some(2))],
+            anchors: &[block_id!(2, "B")],
+            ..Default::default()
+        },
+        TxTemplate {
+            tx_name: "block 2",
+            inputs: &[TxInTemplate::PrevTx("block 1", 0)],
+            outputs: &[TxOutTemplate::new(30000, Some(3))],
+            anchors: &[block_id!(3, "C"), block_id!(4, "D")],
+            ..Default::default()
+        },
+    ];
+    let (graph, _, _) = init_graph(&template);
+    let new_graph = graph.clone().map_anchors(|a| NonDeterministicAnchor {
+        anchor_block: a,
+        // A non-deterministic value
+        non_deterministic_field: rand::thread_rng().next_u32(),
+    });
+
+    // Check all the details in new_graph reconstruct as well
+
+    let mut full_txs_vec: Vec<_> = graph.full_txs().collect();
+    full_txs_vec.sort();
+    let mut new_txs_vec: Vec<_> = new_graph.full_txs().collect();
+    new_txs_vec.sort();
+    let mut new_txs = new_txs_vec.iter();
+
+    for tx_node in full_txs_vec.iter() {
+        let new_txnode = new_txs.next().unwrap();
+        assert_eq!(new_txnode.txid, tx_node.txid);
+        assert_eq!(new_txnode.tx, tx_node.tx);
+        assert_eq!(
+            new_txnode.last_seen_unconfirmed,
+            tx_node.last_seen_unconfirmed
+        );
+        assert_eq!(new_txnode.anchors.len(), tx_node.anchors.len());
+
+        let mut new_anchors = new_txnode.anchors.iter();
+        for anchor in tx_node.anchors.iter() {
+            assert_eq!(new_anchors.next().unwrap().anchor_block, *anchor);
+        }
+        assert!(new_anchors.next().is_none());
+    }
+    assert!(new_txs.next().is_none());
+}
```

Comment:
There are still some checks missing.

* The `TxGraph::anchors` field is not checked. We can use the `TxGraph::all_anchors` method to get it.
* The current test only checks whether all txs in `graph` are in `new_graph`, but it doesn't check for duplicates or potential additions.

Suggestion: If we collect things we want to check into `BTreeSet`, it makes the checks more complete and easier to read.

Reply:
I got it. The all_anchors of a graph are checked, and the tx_node iterator may not follow the block_id order and will be in a different order, so it still checks with a loop.

Reply:
The nature of `BTree{Set|Map}`s is that they are always ordered (hence you can `.collect` and compare the results). So no, comparing inside a loop is not required.

Personally I think this is cleaner, but this is personal preference.